packages/common/src/hooks/useImageSize2.ts

```diff
@@ -0,0 +1,134 @@
+ * @returns The url of the image, or undefined if the image is not available
+ */
+export const useImageSize2 = ({
+  artwork,
+  targetSize,
+  defaultImage
+}: {
+  artwork: Maybe<Track['artwork']>
+  targetSize: SquareSizes
+  defaultImage: string
+}) => {
+  const [imageUrl, setImageUrl] = useState<string | null>(null)
+
+  const preloadImage = (url: string) => {
+    return new Promise((resolve, reject) => {
+      const img = new Image()
+      img.src = url
+      img.onload = () => resolve(url)
+      img.onerror = () => reject(url)
+    })
+  }
+
+  const fetchWithFallback = useCallback(
+    async (url: string) => {
+      const mirrors = [...(artwork?.mirrors ?? [])]
+      let currentUrl = url
+
+      while (mirrors.length > 0) {
+        try {
+          await preloadImage(currentUrl)
+          return currentUrl
+        } catch {
+          const nextMirror = mirrors.shift()
+          if (!nextMirror) throw new Error('No mirror found')
+          const currentHost = new URL(currentUrl).hostname
+          const nextHost = new URL(nextMirror).hostname
+          const newUrl = currentUrl.replace(currentHost, nextHost)
+          currentUrl = newUrl
+        }
+      }
+
+      throw new Error('Failed to fetch image from all mirrors')
+    },
+    [artwork?.mirrors]
+  )
+
+  const resolveImageUrl = useCallback(async () => {
+    if (!artwork) {
+      console.debug(`useImageSize: no artwork, loading`)
+      return
+    }
+    const targetUrl = artwork[targetSize]
+    if (!targetUrl) {
+      console.debug(`useImageSize: no target url for ${targetSize}`)
+      setImageUrl(defaultImage)
+      return
+    }
+
+    if (IMAGE_CACHE.has(targetUrl)) {
+      console.debug(`useImageSize: cache hit ${targetUrl}`)
+      setImageUrl(targetUrl)
+      return
+    }
+
+    // Check for smaller size in the cache
+    // If found, set the image url and preload the actual target url
+    const smallerSize = Object.keys(artwork).find(
+      (size) =>
+        parseInt(size) < parseInt(targetSize) &&
+        artwork[size as SquareSizes] &&
+        IMAGE_CACHE.has(artwork[size as SquareSizes]!)
+    ) as SquareSizes | undefined
+
+    // Check for larger size
+    // If found, set the image url to the larger size and return
+    const largerSize = Object.keys(artwork).find(
+      (size) =>
+        parseInt(size) > parseInt(targetSize) &&
+        artwork[size as SquareSizes] &&
+        IMAGE_CACHE.has(artwork[size as SquareSizes]!)
+    ) as SquareSizes | undefined
+
+    if (largerSize) {
+      console.debug(
+        `useImageSize: cache miss, found larger ${targetUrl} ${largerSize}`
+      )
+      setImageUrl(artwork[largerSize] ?? null)
+      return
+    }
+
+    if (smallerSize) {
+      console.debug(
+        `useImageSize: cache miss, found smaller ${targetUrl} ${smallerSize}`
+      )
+      setImageUrl(artwork[smallerSize] ?? null)
+      const finalUrl = await fetchWithFallback(targetUrl)
```

Comment:
It occurs to me that while we have logic to short circuit if we find a cached image, the images are only added to the cache after they've been fetched. If I were to use 5 instances of this hook at once, would they all initiate parallel fetches?

Reply:
yeah. could try to get fancy and solve this but i think it's a somewhat unlikely scenario and the browser might also be smart enough to dedupe. our existing code has this same problem PLUS it's using redux so even more likely to happen. can make a tix for this though